File: YCFirstTime.m
//
//  YCFirstTime.m
//  YU - YUPPIU
//
//  Created by Fabio Knoedt on 28/08/14.
//  Copyright (c) 2014 Fabio Knoedt. All rights reserved.
//

#import "YCFirstTime.h"
#import "YCFirstTimeObject.h"

/// WARNING: don't change that or you will lose information for the next session.
#define sharedGroup     @"sharedGroup"

@interface YCFirstTime ()

/*!
 *  The NSMutableDictionary to persist the block execution history.
 */
@property (nonatomic, retain) NSMutableDictionary *fkDict;

@end

/*!
 *  Implementation for YCFirstTime.
 */
@implementation YCFirstTime

#pragma mark - Init

/*!
 *  Implemented by subclasses to initialize a new object (the receiver) immediately after memory for it has been allocated.
 *  @return An initialized object, or nil if an object could not be created for some reason that would not result in an exception.
 */
- (instancetype)init;
{
    if ((self = [super init])) {
        
        /// Load the already tracked executed blocks from UserDefaults.
        self.fkDict = [self loadDictionaryFromUserDefaults];
    }
    
    return self;
}

/*!
 *  Singleton instance shared for the app session.
 *  @return the initialized object.
 */
+ (YCFirstTime *)shared;
{
    static YCFirstTime *sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[YCFirstTime alloc] init];
    });
    return sharedInstance;
}

#pragma mark - Public methods

/*!
 *  Execute a block only once.
 *  @param blockOnce    The block to be executed.
 *  @param blockKey     The unique name of the block.
 */
- (void)executeOnce:(void (^)())blockOnce
             forKey:(NSString *)blockKey;
{
    [self executeOnce:blockOnce executeAfterFirstTime:nil forKey:blockKey perVersion:FALSE everyXDays:0];
}

/*!
 *  Execute a block only once.
 *  @param  blockOnce            The block to be executed only once.
 *  @param  blockAfterFirstTime  The block to be executed always.
 *  @param  blockKey             The unique name of the block.
 */
- (void)executeOnce:(void (^)())blockOnce executeAfterFirstTime:(void (^)())blockAfterFirstTime
             forKey:(NSString *)blockKey;
{
    [self executeOnce:blockOnce executeAfterFirstTime:blockAfterFirstTime forKey:blockKey perVersion:FALSE everyXDays:0];
}

/*!
 *  Execute a block only once per version.
 *  @param  blockOnce            The block to be executed only once.
 *  @param  blockKey             The unique name of the block.
 */
- (void)executeOncePerVersion:(void (^)())blockOnce
                       forKey:(NSString *)blockKey;
{
    [self executeOnce:blockOnce executeAfterFirstTime:nil forKey:blockKey perVersion:TRUE everyXDays:0];
}

/*!
 *  Execute a block only once per version.
 *  @param  blockOnce            The block to be executed only once.
 *  @param  blockAfterFirstTime  The block to be executed always.
 *  @param  blockKey             The unique name of the block.
 */
- (void)executeOncePerVersion:(void (^)())blockOnce
        executeAfterFirstTime:(void (^)())blockAfterFirstTime
                       forKey:(NSString *)blockKey;
{
    [self executeOnce:blockOnce executeAfterFirstTime:blockAfterFirstTime forKey:blockKey perVersion:TRUE everyXDays:0];
}

/*!
 *  Execute a block only once.
 *  @param  blockOnce            The block to be executed only once.
 *  @param  blockKey             The unique name of the block.
 *  @param  days                 The number of days that the code should be executed again.
 */
- (void)executeOncePerInterval:(void (^)())blockOnce
                        forKey:(NSString *)blockKey
              withDaysInterval:(float)days;
{
    [self executeOnce:blockOnce executeAfterFirstTime:nil forKey:blockKey perVersion:FALSE everyXDays:days];
}

/*!
 *  Execute a block only once.
 *  @param  blockOnce            The block to be executed only once.
 *  @param  blockAfterFirstTime  The block to be executed always.
 *  @param  blockKey             The unique name of the block.
 *  @param  checkVersion         Execute this block every new version.
 */
- (void)executeOnce:(void (^)())blockOnce executeAfterFirstTime:(void (^)())blockAfterFirstTime forKey:(NSString *)blockKey perVersion:(BOOL)checkVersion everyXDays:(float)days;
{
    /// Check if the block was executed already.
    if ([self blockAlreadyExecutedForKey:blockKey perVersion:checkVersion everyXDays:days]) {

        /// Execute the blockAfterFirstTime from the second time on.
        if (blockAfterFirstTime) {
            blockAfterFirstTime();
        }
        
    } else {
        
        /// If there is a valid block.
        if (blockOnce) {
            
            /// Execute block.
            blockOnce();
            
            /// Save execution information.
            [self saveExecutionInformationForKey:blockKey forGroup:sharedGroup];
        }
    }
}

/*!
 *  Check if a block was executed already or not.
 *  @param blockKey The unique name of the block.
 */
- (BOOL)blockWasExecuted:(NSString *)blockKey;
{
    return [self blockAlreadyExecutedForKey:blockKey perVersion:FALSE everyXDays:0];
}

#pragma mark - Private methods

/*!
 *  Check if a block was executed already.
 *  @param blockKey     The unique name of the block.
 *  @param checkVersion If the block should be executed every new version or not.
 *  @return a boolean if the block was executed or not.
 */
- (BOOL)blockAlreadyExecutedForKey:(NSString *)blockKey perVersion:(BOOL)checkVersion everyXDays:(float)days;
{
    /// Boolean for executed blocks.
    BOOL executed = FALSE;
    
    /// Get the key dictionary.
    YCFirstTimeObject *blockInfo = [self getInfoForBlock:blockKey forGroup:sharedGroup];
    if (blockInfo) {
        
        /// Boolean for executed blocks.
        executed = TRUE;
    }
    
    /// Version.
    if (checkVersion && blockInfo) {
        
        NSString *currentVersion = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleShortVersionString"];
        executed = executed && [currentVersion isEqualToString:blockInfo.lastVersion];
    }
    
    /// Every X days.
    if (days && blockInfo) {
        
        float differenceInSeconds = [[NSDate date] timeIntervalSinceDate:blockInfo.lastTime];
        executed = executed && differenceInSeconds/84600 < days;
    }
    
    return executed;
}

/*!
 *  Saves the execution of a block. It saves on the disk to post check.
 *  @param  blockKey The unique name of the block.
 *  @param  groupKey The unique name of the group block.
 */
- (void)saveExecutionInformationForKey:(NSString *)blockKey
                              forGroup:(NSString *)groupKey;
{
    YCFirstTimeObject *blockInfo = [self getInfoForBlock:blockKey forGroup:groupKey];
    if (!blockInfo) {
        blockInfo = [[YCFirstTimeObject alloc] init];
    }
    
    /// Set the version.
    blockInfo.lastVersion = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleShortVersionString"];
    
    /// Set last time execute date.
    blockInfo.lastTime = [NSDate date];
    
    /// Set to the main dictionary.
    [self setInfoForBlock:blockInfo forKey:blockKey forGroup:groupKey];
}

/*!
 *  Get the info for a certain block key and group.
 *  @param  blockKey     The unique name of the block.
 *  @param  groupKey     The unique group name that this block is part of.
 *  @return the YCFirstTimeObject with the block execution information.
 */
- (YCFirstTimeObject *)getInfoForBlock:(NSString *)blockKey forGroup:(NSString *)groupKey;
{
    return [[self getDictForGroup:groupKey] objectForKey:blockKey];
}

/*!
 *  Get the NSMutableDictionary for a certain block group.
 *  @param  groupKey     The unique group name that this block is part of.
 *  @return The NSMutableDictionary for the group.
 */
- (NSMutableDictionary *)getDictForGroup:(NSString *)groupKey;
{
    if ([groupKey length]) {
        return [_fkDict objectForKey:groupKey];
    } else {
        return [_fkDict objectForKey:sharedGroup];
    }
}

/*!
 *  Set the info for a certain block key and group.
 *  @param  blockInfo    The block information to be saved.
 *  @param  blockKey     The unique name of the block.
 *  @param  groupKey     The unique group name that this block is part of.
 */
- (void)setInfoForBlock:(YCFirstTimeObject *)blockInfo forKey:(NSString *)blockKey forGroup:(NSString *)groupKey;
{
    /// Get the right group dictionary (specific or general).
    NSMutableDictionary *groupDictionary = [self getDictForGroup:groupKey];
    if (!groupDictionary) {
        groupDictionary = [NSMutableDictionary dictionary];
    }
    
    /// Set the block info for the dictionary.
    [groupDictionary setObject:blockInfo forKey:blockKey];
    if (!_fkDict) {
        _fkDict = [NSMutableDictionary dictionary];
    }
    [_fkDict setObject:groupDictionary forKey:[groupKey length] ? groupKey : sharedGroup];
    
    /// Sync with the disk.
    [self saveDictionaryToUserDefaults];
}

#pragma mark - UserDefaults

/*!
 *  Resets/Erases all the previous executions.
 */
- (void)reset;
{
    /// Delete the values for the sharedKey.
    [[NSUserDefaults standardUserDefaults] removeObjectForKey:sharedGroup];

    /// Loads again from userDefaults.
    _fkDict = nil;
}

/*!
 *  Loads the dictionary from User Defaults.
 *  @return the saved dictionary.
 */
- (NSMutableDictionary *)loadDictionaryFromUserDefaults;
{
    /// Load the encoded dictionary from User Defaults and decode it.
    NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
    NSData *encodedDictionary = [userDefaults objectForKey:NSStringFromClass([self class])];
    if (encodedDictionary) {
        return [NSKeyedUnarchiver unarchiveObjectWithData:encodedDictionary];
    } else {
        return [NSMutableDictionary dictionary];
    }
}

/*!
 *  Encodes and saves the current dictionary to the User Defaults.
 */
- (void)saveDictionaryToUserDefaults;
{
    /// Encodes and Saves the decoded running dictionary to User Defaults.
    NSData *decodedDictionary = [NSKeyedArchiver archivedDataWithRootObject:self.fkDict];
    [[NSUserDefaults standardUserDefaults] setObject:decodedDictionary forKey:NSStringFromClass([self class])];
    [[NSUserDefaults standardUserDefaults] synchronize];
}

@end

File: YCFirstTime.swift
//
//  YCFirstTime.swift
//
//  Swift port of YCFirstTime.{h,m}. Keeps @objc exports under the original
//  class name and selectors, and preserves the on-disk archive contract:
//
//    UserDefaults key: "YCFirstTime"
//    NSKeyedArchiver shape: { "sharedGroup": { blockKey: YCFirstTimeObject } }
//
//  Two long-standing bugs from the Obj-C version are fixed here (see the
//  companion test commit):
//    - `/84600` typo in the day math is now `/86400`.
//    - `reset()` previously removed the wrong UserDefaults key
//      ("sharedGroup"); it now removes the archive key.
//

import Foundation

// The sharedGroup value that the Obj-C implementation wrote into every
// persisted archive. Must stay identical so existing on-disk state decodes.
private let kSharedGroup = "sharedGroup"

// The UserDefaults key under which the archive lives. Matches
// `NSStringFromClass([YCFirstTime class])` used by the Obj-C version.
private let kDefaultsKey = "YCFirstTime"

private let kSecondsPerDay: TimeInterval = 86_400

@objc(YCFirstTime)
public final class YCFirstTime: NSObject {

    private var fkDict: NSMutableDictionary

    // Injectable seams. Default providers match the original behavior
    // (CFBundleShortVersionString / Date()). Tests override via the
    // Swift-visible setters below — also exported to Obj-C so the existing
    // test bridging header keeps working unchanged.
    @objc public var versionProvider: (() -> String?)?
    @objc public var nowProvider: (() -> Date)?

    @objc public static let shared = YCFirstTime()

    @objc public override init() {
        self.fkDict = Self.loadDictionary()
        super.init()
        self.versionProvider = {
            Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String
        }
        self.nowProvider = { Date() }
    }

    // MARK: - Public API (mirrors the Obj-C selectors)

    @objc(executeOnce:forKey:)
    public func executeOnce(_ block: (() -> Void)?, forKey key: String) {
        execute(block, afterFirstTime: nil, forKey: key, perVersion: false, everyXDays: 0)
    }

    @objc(executeOnce:executeAfterFirstTime:forKey:)
    public func executeOnce(
        _ block: (() -> Void)?,
        executeAfterFirstTime afterBlock: (() -> Void)?,
        forKey key: String
    ) {
        execute(block, afterFirstTime: afterBlock, forKey: key, perVersion: false, everyXDays: 0)
    }

    @objc(executeOncePerVersion:forKey:)
    public func executeOncePerVersion(_ block: (() -> Void)?, forKey key: String) {
        execute(block, afterFirstTime: nil, forKey: key, perVersion: true, everyXDays: 0)
    }

    @objc(executeOncePerVersion:executeAfterFirstTime:forKey:)
    public func executeOncePerVersion(
        _ block: (() -> Void)?,
        executeAfterFirstTime afterBlock: (() -> Void)?,
        forKey key: String
    ) {
        execute(block, afterFirstTime: afterBlock, forKey: key, perVersion: true, everyXDays: 0)
    }

    @objc(executeOncePerInterval:forKey:withDaysInterval:)
    public func executeOncePerInterval(
        _ block: (() -> Void)?,
        forKey key: String,
        withDaysInterval days: Float
    ) {
        execute(block, afterFirstTime: nil, forKey: key, perVersion: false, everyXDays: days)
    }

    @objc(blockWasExecuted:)
    public func blockWasExecuted(_ key: String) -> Bool {
        alreadyExecuted(forKey: key, perVersion: false, everyXDays: 0)
    }

    @objc public func reset() {
        UserDefaults.standard.removeObject(forKey: kDefaultsKey)
        fkDict = NSMutableDictionary()
    }

    // MARK: - Core logic

    private func execute(
        _ block: (() -> Void)?,
        afterFirstTime afterBlock: (() -> Void)?,
        forKey key: String,
        perVersion: Bool,
        everyXDays days: Float
    ) {
        if alreadyExecuted(forKey: key, perVersion: perVersion, everyXDays: days) {
            afterBlock?()
            return
        }
        guard let block else { return }
        block()
        saveExecution(forKey: key, forGroup: kSharedGroup)
    }

    private func alreadyExecuted(
        forKey key: String,
        perVersion: Bool,
        everyXDays days: Float
    ) -> Bool {
        guard let info = info(forBlock: key, forGroup: kSharedGroup) else { return false }

        if perVersion {
            let currentVersion = versionProvider?() ?? Self.defaultVersion()
            if currentVersion != info.lastVersion { return false }
        }

        if days > 0, let lastTime = info.lastTime {
            let now = nowProvider?() ?? Date()
            let elapsed = now.timeIntervalSince(lastTime)
            // Fixes the /84600 typo from the Obj-C version (YCFirstTime.m:194).
            if elapsed / kSecondsPerDay >= Double(days) { return false }
        }

        return true
    }

    private func saveExecution(forKey key: String, forGroup group: String) {
        let info = info(forBlock: key, forGroup: group) ?? YCFirstTimeObject()
        info.lastVersion = versionProvider?() ?? Self.defaultVersion()
        info.lastTime = nowProvider?() ?? Date()
        setInfo(info, forBlock: key, forGroup: group)
    }

    private func info(forBlock key: String, forGroup group: String) -> YCFirstTimeObject? {
        let resolvedGroup = group.isEmpty ? kSharedGroup : group
        let groupDict = fkDict[resolvedGroup] as? NSDictionary
        return groupDict?[key] as? YCFirstTimeObject
    }

    private func setInfo(_ info: YCFirstTimeObject, forBlock key: String, forGroup group: String) {
        let resolvedGroup = group.isEmpty ? kSharedGroup : group
        let groupDict = (fkDict[resolvedGroup] as? NSMutableDictionary) ?? NSMutableDictionary()
        groupDict[key] = info
        fkDict[resolvedGroup] = groupDict
        saveDictionary()
    }

    // MARK: - Persistence

    private static func defaultVersion() -> String? {
        Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String
    }

    private static func loadDictionary() -> NSMutableDictionary {
        guard let data = UserDefaults.standard.data(forKey: kDefaultsKey) else {
            return NSMutableDictionary()
        }
        let allowed: [AnyClass] = [
            NSMutableDictionary.self, NSDictionary.self,
            NSString.self, NSDate.self, YCFirstTimeObject.self,
        ]
        let decoded = try? NSKeyedUnarchiver.unarchivedObject(ofClasses: allowed, from: data)
        return (decoded as? NSMutableDictionary) ?? NSMutableDictionary()
    }

    private func saveDictionary() {
        guard let data = try? NSKeyedArchiver.archivedData(
            withRootObject: fkDict,
            requiringSecureCoding: true
        ) else { return }
        UserDefaults.standard.set(data, forKey: kDefaultsKey)
    }
}
